### ISSUE TITLE
VS Code freezes regularly on macOS 12.3 with Apple M1

### ISSUE THREAD

**Issue by stefanbschneider** [microsoft/vscode#145387]
I'm using the latest stable VS Code on my MacBook Pro 2021 with M1 chip for programming Python (and C++).
I have been having issues with VS Code randomly slowing down and freezing until macOS asks to stop it. Then I have to terminate VS Code and restart it. That happens multiple times within a couple of hours, so it's really annoying when coding & debugging.

The freeze happens out of nowhere. CPU load is very low, I'm not doing anything special or running expensive other programs in the background.
I can't really reproduce the freeze quickly, but it does happen regularly (roughly every 0,5-2h).

Seems like other users also have this issue: https://github.com/microsoft/vscode/issues/122111 and https://github.com/microsoft/vscode/issues/129076
Unfortunately, these issues were closed and locked without providing a solution.

- VS Code Version: 1.65.2
- OS Version: macOS 12.3
- MacBook Pro 2021 with Apple M1 Pro chip

My extensions are: 
```
code --install-extension alefragnani.Bookmarks
code --install-extension cschlosser.doxdocgen
code --install-extension eamodio.gitlens
code --install-extension jeff-hykin.better-cpp-syntax
code --install-extension mhutchie.git-graph
code --install-extension ms-python.python
code --install-extension ms-python.vscode-pylance
code --install-extension ms-toolsai.jupyter
code --install-extension ms-toolsai.jupyter-keymap
code --install-extension ms-toolsai.jupyter-renderers
code --install-extension ms-vscode-remote.remote-containers
code --install-extension ms-vscode-remote.remote-ssh
code --install-extension ms-vscode-remote.remote-ssh-edit
code --install-extension ms-vscode-remote.remote-wsl
code --install-extension ms-vscode.cmake-tools
code --install-extension ms-vscode.cpptools
code --install-extension ms-vscode.cpptools-extension-pack
code --install-extension ms-vscode.cpptools-themes
code --install-extension njpwerner.autodocstring
code --install-extension trond-snekvik.simple-rst
code --install-extension twxs.cmake
```
Since I need the extensions for development and I can't quickly reproduce the freeze, I couldn't test if the freeze eventually occurs with all extensions disabled.

Also, I tried recording a crash dump as described [here](https://github.com/microsoft/vscode/wiki/Native-Crash-Issues), but no dump was created. Maybe because VS Code froze and didn't crash.

Any help is really appreciated. This really disturbs my coding workflow.

**Comment by stefanbschneider** [comment]
Any ideas? My guess is that this is connected to mypy somehow.

**Comment by deepak1556** [comment]
Can you reproduce the freeze with `1.66` which comes with a newer runtime ? If it still repros, can you take a process dump following the steps below,

```
* Once the freeze happens, note the process ID of the window/workbench that froze using `code --status`
* Open Activity monitor and find the process with the same ID
* Double click on the process and choose the sample option button in the new window
* Attach the generated sample in this issue
```

**Comment by deepak1556** [comment]
Also are you using the apple silicon build or the universal build ?

**Comment by deepak1556** [comment]
Please take process dump for the window process which has the high CPU consumption.

**Comment by seanbischoff** [comment]
@deepak1556 I'd closed my ticket - I'm on an Intel Mac - and it is still doing the same thing. I cannot get you a sample because the process is not responding - the sample window just hangs. I can't even screen capture it because nothing responds now.

**Comment by ConnorDoesDev** [comment]
I'm also having this issue on a MacBook Air M1 Monterey (12.1) - VS Code version 1.66.0

**Comment by ConnorDoesDev** [comment]
Update: I've downloaded [Visual Studio Code for Insiders](https://code.visualstudio.com/insiders) and it seems to work fine.

**Comment by seanbischoff** [comment]
> 

Our team suspects our use of iTerm and NodeJS. I'm planning to work with these & another IDE, and see if my Mac still freezes today.

**Comment by seanbischoff** [comment]
Happened with the other IDE, instant freeze, fan spunup & laptop crash/restart. 5 minutes after starting up local server, Node.js v14.15.4.

**Comment by sumerc** [comment]
Same issue. Regularly getting. Except in my case it never happens when docker is not running. It seems to be related with docker. Will try getting a dump.

**Comment by sumerc** [comment]
Problem still persists. It happens with `1.67.2` but not with `1.67.0` insiders version. This is an incredibly annoying problem and I would love to help for anyone interested in debugging this issue...

**Comment by trkee** [comment]
+1

**Comment by nikhilsarvaiye** [comment]
facing same issue

**Comment by wtclarke** [comment]
Also having this issue on an Intel mac, macos 12.1. Occured after I updated from pylance v2021 to v2022.

**Comment by trkee** [comment]
I found that excluding vscode from the time machine seems to solve the
problem. At least for about a week now it has been way better.

**Comment by wtclarke** [comment]
@trkee As in the workspace folder or the actual vscode app?

**Comment by trkee** [comment]
https://support.apple.com/en-il/guide/mac-help/mh15622/mac

**Comment by trkee** [comment]
The vscode app.

**Comment by seanbischoff** [comment]
For me, either slow freezes or hard fast crashes with VSCode, both stable & Insiders edition, and Webstorm, _when running latest version of Node w/Nginx_. We suspect it's with the OS memory management.

**Comment by glaide** [comment]
I'm facing the same issue here, with my mac M1, my OS is Monterey 12.3.1, vscode version: 1.67.2. I use vscode to run flutter, and when I use Android Studio it works pretty well.

**Comment by codefingas** [comment]
Facing the same issue. ☹️

**Comment by deepak1556** [comment]
@sumerc thanks for the `Code Helper (Renderer)` process sample, can you also provide the sample of `Code Helper` and `Code` processes when the freeze happens. Also can you check the following,

(i) Does launching with `code --disable-features="ParkableStringsToDisk"` prevents the hang
(ii) Does the issue happen with latest insiders https://code.visualstudio.com/insiders that comes with a newer runtime.

**Comment by sumerc** [comment]
> Does the issue happen with latest insiders https://code.visualstudio.com/insiders that comes with a newer runtime.

I have been using Insiders `1.67.0` for a while now without any problem. I get this issue only when I use the normal version. Just to make sure: even I can reproduce the issue on `1.68.0`. 

I will try to get the information you requested.

**Comment by sumerc** [comment]
I have reproduced the problem again. I ran the `code` with `--disable-features="ParkableStringsToDisk"` but get the warning above, so unsure if the flag has an effect or not.

I get samples from the following processes:

1. Code
2. Code Helper (There were several of them, I just selected one, no one seems to be using more CPU)
3. Code Helper GPU (not too much CPU)
4. Code Helper (Renderer) (There were several of them, I just selected the one that has %100 all the time)

[Sample of Code Helper (GPU)2.txt](https://github.com/microsoft/vscode/files/8907469/Sample.of.Code.Helper.GPU.2.txt)
[Sample of Code Helper2.txt](https://github.com/microsoft/vscode/files/8907470/Sample.of.Code.Helper2.txt)
[Sample of Code2.txt](https://github.com/microsoft/vscode/files/8907471/Sample.of.Code2.txt)
[Sample of Code Helper (Renderer)2.txt](https://github.com/microsoft/vscode/files/8907472/Sample.of.Code.Helper.Renderer.2.txt)

**Comment by deepak1556** [comment]
Thanks for the logs, the warning is just to convey that the flag will be passed down to the runtime, you can ignore it.

From the `Code Helper (Renderer)` sample most of the cycle is spent on JIT code that cannot be symbolicated. As a next step, can you take a JS profile following the steps below,

```
* Launch vscode stable normally
* Open Command Palette and run `Toggle Developer Tools`
* Detach the devtools from the workbench into a separate window
* Choose Javascript Profiler tab from the devtools
* Start recording
* Once the workbench freezes
* Stop the profiler and attach the generated trace
```

> I have been using Insiders 1.67.0 for a while now without any problem. I get this issue only when I use the normal version. Do you still want me to try the newest Insider version?

Is there a difference in extensions or settings used between stable and insiders ? Does the issue repro with insiders if you try with the same profile as stable.

**Comment by sumerc** [comment]
> Is there a difference in extensions or settings used between stable and insiders ? Does the issue repro with insiders if you try with the same profile as stable.

Yes, the extension were different and I do a quick check: I uninstalled the ones (5 in total) there were not installed in the Insiders and the issue did not reproduced. So, I will (in parallel) try to find the problematic extension.

> sample most of the cycle is spent on JIT code that cannot be symbolicated. As a next step, can you take a JS profile following the steps below,

Nice to know where the unsymbolicated frames were coming from. I was curious. Anyway, I have done that and reproduced the problem. However, the issue is: the resulting file is `159Mb` and as I am using a company laptop, I am really unsure about what the profile can contain, I might need to ask security team. Having said that, is there anything you would like to see in the CPU profile? Below is a small screenshot of the top functions:

<img width="721" alt="Screen Shot 2022-06-16 at 20 59 26" src="https://user-images.githubusercontent.com/136531/174136007-110b5651-d6a6-42da-bced-82ddf827e7e3.png">

BTW, I am not closing the stuck app. right now just in case you would need to request any information by attaching it somehow?

**Comment by sumerc** [comment]
@deepak1556, another point: title is misleading that this issue only happens on `macOS 12.3`, however I can reproduce it in `12.4` as well.

**Comment by sumerc** [comment]
I am writing for future reference. I have disabled `CMake` and `CMake tools` extensions and the problem is fixed for me. So, I believe there is some extension API that causes the renderer process to get stuck. @deepak1556 , is there any other way to debug this issue other than the javascript profile? Do you still think you need the JS profile to move further on this issue?

**Comment by deepak1556** [comment]
@sumerc sorry for the delay, tackling some milestone related issues. If you can share the JS profile that would be helpful to move the issue forward, also there shouldn't be any private data that should be of concern in sharing the profile. The call stacks should only be from VSCode core and it extensions. If there is concern, you can share the profile to my contact Deepak.Mohan@microsoft.com

Since you have narrowed down the problematic extensions, can you list any specific steps to be followed when recreating this issue (ex: install said extension and open a particular type of file) ? If possible, I would like to repro the issue on my setup which can also help simplify the debugging part.

**Comment by sumerc** [comment]
> If there is concern, you can share the profile to my contact [Deepak.Mohan@microsoft.com](mailto:Deepak.Mohan@microsoft.com)

Ok. I will reproduce it again(without involving any repos) and send you the profile.

> Since you have narrowed down the problematic extensions, can you list any specific steps to be followed when recreating this issue (ex: install said extension and open a particular type of file) ? If possible, I would like to repro the issue on my setup which can also help simplify the debugging part.

For this, I will need to find out if the extensions are the root cause the problem or the extensions are conflicting with some other extension? In order to do that, I will try to reproduce the issue in a separate VS Code profile and see what happens.

**Comment by Rickion** [comment]
Same issue. And I found something strange. 
When `VS Code` is not responding, my Chrome costs CPU over 50%:
![image](https://user-images.githubusercontent.com/13317720/176412404-e8d1c48e-0514-4b5a-8380-a0b7554efb18.png)
![image](https://user-images.githubusercontent.com/13317720/176412634-853de457-1aed-4776-b155-972dc210d345.png)

After a few secs:
![image](https://user-images.githubusercontent.com/13317720/176414318-8c7a06f6-dc8b-43b5-8773-117d9ddcb1ff.png)


Only `VS Code` is not responding. Other Apps works well. After 10-20 sec, everything turns to good.
This happens regularly when I switch apps bewteen `chrome` and `vscode`.

Run `code -status` seems good: 
```
Version:          Code 1.68.1 (30d9c6cd9483b2cc586687151bcbcd635f373630, 2022-06-14T12:52:13.188Z)
OS Version:       Darwin x64 21.5.0
CPUs:             Apple M1 (8 x 24)
Memory (System):  16.00GB (0.04GB free)
Load (avg):       2, 1, 2
VM:               0%
Screen Reader:    no
Process Argv:     --crash-reporter-id 76ce2e8d-304c-4701-abd6-0229d6cd3a3a
GPU Status:       2d_canvas:                              enabled
                  canvas_oop_rasterization:               disabled_off
                  direct_rendering_display_compositor:    disabled_off_ok
                  gpu_compositing:                        enabled
                  metal:                                  disabled_off
                  multiple_raster_threads:                enabled_on
                  oop_rasterization:                      enabled
                  opengl:                                 enabled_on
                  rasterization:                          enabled
                  raw_draw:                               disabled_off_ok
                  skia_renderer:                          enabled_on
                  video_decode:                           enabled
                  video_encode:                           enabled
                  webgl:                                  enabled
                  webgl2:                                 enabled

CPU %	Mem MB	   PID	Process
   10	    66	   994	code main
    0	    49	   997	   gpu-process
    0	    16	  1000	   utility-network-service
    0	    33	  1001	   window (xxxx.py — xxx9)
    0	   180	  1004	   window (Content.js — XXX)
    0	    33	  1534	   shared-process
    0	    16	  1547	     ptyHost
    0	     0	  2259	       /bin/zsh -l
    0	     0	  2430	       /bin/zsh -l
    0	     0	 75606	       /bin/zsh -l
    0	     0	 96607	       /bin/zsh -l
    0	    16	  2054	     fileWatcher
    0	    16	 60364	     fileWatcher
    0	     0	 96606	     fileWatcher
    0	    16	  1544	   extensionHost
    0	    16	  4718	     /Applications/APPs/Visual Studio Code.app/
```

Update chrome not help. (Current version 103.0.5060.53 )

**Comment by realnatuki** [comment]
I just had a complete macos crash on M1 when opening vscode. 

Vscode opened and  It popped a recommendation for an angular extension,  after clicking on extension it first froze my mouse and keyboard and then the mac restarted. After it restarted I got a send report to apple popup,  by clicking on send report it then also froze the whole computer. Not sure if it's vscode or M1 related issue. IMHO M1 is the worst computer experience i've ever had  including windows 98 and windows vista. Vscode on my old intel mac has never crashed and is still working fine.

**Comment by realnatuki** [comment]
This is the crash report i had. Doesn't even seem related to vscode it's showing logs about xcode ios coresimulator. I use vscode for python and angular development no simulator required and xcode wasn't even running.  

<img width="1154" alt="Screenshot 2022-07-03 at 01 08 12" src="https://user-images.githubusercontent.com/85006141/177019569-0f97af1f-b45d-4c22-8aed-cd04acafffec.png">

**Comment by thedeterminedsoul** [comment]
I have this issue on both intel and M1 macs. All of a sudden visual studio code freezes and slowly it freezes entire computer. +1 for the issue.

**Comment by shennan** [comment]
I get this issue (same behaviour but perhaps different reasons to OP) when I run a bash script for deployment which does a lot of cleaning and building of js packages in a monorepo/workspace that I currently have open in VS Code. I also get it sometimes when just doing a `npx npkill` to dump all `node_module` folders in the monorepo, sometimes followed by a `yarn install`. The commands don't error out, but often I'll have to restart VS Code.

Same behaviour; VS Code freezes and after a while it either recovers or gives me an unresponsive modal. Most of the time I have to force quite the app and restart.

I'm using macOS Monterey 12.4 on an M1 Pro.

**Comment by michaelw76** [comment]
Hello, I have been experiencing more or less the same issue as folks were experiencing here - except that my VS Code would freeze the moment I would click on the "Open" link in the "Get Started" tab upon starting up VS Code.  What I have done is I have installed Xcode Version 14.0 beta 3 (14A5270f) as well as the CLI tools for it.  What I have than done is install Visual Studio for Mac version 17.0 - I made sure to check the option during installation to download the libraries for MacOS development.  Once installation finished, I have opened up Visual Studio for Mac - it worked fine.  I than deleted Visual Studio for Mac and installed VS Code from this link:  https://code.visualstudio.com/?wt.mc_id=DX_841432 i.e. version 1.69.1 (universal).  And it now works like a charm - without freezing.

Unfortunately I have not tested whether this issue got resolved due to me installing Xcode, or installing Visual Studio for Mac - but I would suggest trying Xcode first, than VS Code - and if it doesn't kill the freezing, delete VS Code, install Visual Studio for Mac, than VSCode and see if that helps.

I am running this Mac:

macOS Monterey Version 12.4
MacBook Air (M1, 2020)
Chip:  Apple M1
Memory:  8Gb

**Comment by mmalik23** [comment]
I was having this issue. I fixed it by removing Sophos Home AV. It occured when compiling code through VSCode, not when compiling through another IDE or the command line. 

My entire computer would also freeze when navigating to classes ( cpu/mem usage was be normal). I think the AV was either preventing the language server from properly reading the build files or blocking the connection between the text editor and the language server which uses rpc ( there were some connection errors in the logs)

I am using VSCode for scala development, so my setup is quite different from users here, however, this worked for me so I am putting this out there. 

VScode 1.6.9

Extensions: Metals, scala-lang

macOS Monterey Version 12.4
MacBook Pro ( 14", M1, 2021)
CPU: 10 cores
Memory: 32Gb

**Comment by sumerc** [comment]
For future reference and FYI @deepak1556: 

I have removed CMake and CMake tool extensions and my freeze problem went away.

![image](https://user-images.githubusercontent.com/136531/180729824-79972679-a594-4d21-8c42-db30f9a62745.png)

I think this is a hard issue to reproduce and debug, I suggest anyone facing this issue to disable some part of their extensions to see what is causing the problem and report it here. For me, it was `CMake` however I am still suspecting that there might be a problematic API that is called from the extension that causes the problem.

**Comment by daniel-han-ioof** [comment]
same issue here on M1 mac 12.5

**Comment by VAIMORSS** [comment]
Getting the same issue with 12.0.1 (21A559), especially when changes are made in the package.json file (React project)

**Comment by exxetaweiss** [comment]
Same here with M1 Pro, 12.5 and VS Code 1.69.2.
This looks like a correlation with the SSD, because when the VS code hangs, the whole system hangs.

**Comment by geekmong** [comment]
same issue here on M1 mac 12.5

**Comment by guiruiz** [comment]
same issue here on M1 mac 12.5

**Comment by thanhson1994** [comment]
Same issue with M1 Pro 14inch and VS code 1.70.2

**Comment by sri-k-stack** [comment]
any updates on this? Same here with 1.64.2 (Universal) and M1

**Comment by ppamorim** [comment]
Spinning ball all the time for me while editing a tiny txt file, what is happening?

**Comment by CorpusCallosum** [comment]
+1 froze for me today, could not force quit. Had to force shutdown my Mac.

**Comment by peteryev** [comment]
+1

**Comment by schellenbergk** [comment]
+1

**Comment by VSCodeTriageBot** [comment]
This issue has been closed automatically because it needs more information and has not had recent activity. See also our [issue reporting](https://aka.ms/vscodeissuereporting) guidelines.

Happy Coding!

**Comment by gvsakki** [comment]
I am still facing same issue with M2 chip

**Comment by daveelton** [comment]
Potential solution: 
Ensure that VsCode is running from an internal drive.

VsCode was running from an external drive. In my case I'd coppied it there, but this is easily possible if you are running it from a browser downloads folder that is on an external drive.

My Mac Mini's tb ports seem to be briefly disconnecting the external drive every so often. This is a known issue with the mac and other apps (such as Dropbox which, terifyingly, treats the disconnection as a mass delete event)

Electron, doesn't handle this disconnection well either and locks up to the point where even force quit isn't reliable.

HTH

**Comment by Linchin** [comment]
I am also facing this issue, though I'm using an intel chip. It started about two weeks ago.

**Comment by msayed-net** [comment]
I'm facing same issue

**Comment by AlexKarpov98** [comment]
I'm facing same issue

**Comment by deepak1556** [comment]
@Linchin @msayed-net @AlexKarpov98 please follow the steps mentioned in https://github.com/microsoft/vscode/issues/145387#issuecomment-1340707320 and open a new issue with the generated logs, thanks!

**Comment by tamuratak** [comment]
Reported at microsoft/vscode-cmake-tools/issues/2967
